### ISSUE TITLE
[RFE] sd-boot: Flickerfree

### ISSUE THREAD

**Issue by AdrianVovk** [systemd/systemd#11693]
Hello!

I'm working towards flickerfree boot and the preservation of the BGRT throughout the boot process in my distro. There's a period between (what seems like) the firmware and the kernel where the screen gets cleared. A quick glance through the code seems to indicate that sd-boot clears the screen and then waits for the user to bring up the menu.

Can this get reversed? sd-boot can clear the screen if it actually needs to print something.

Thanks!

**Comment by yuwata** [comment]
Maybe duplicate of #10666?

**Comment by AdrianVovk** [comment]
@yuwata I set the timeout to 0 and it still flickers

**Comment by poettering** [comment]
We only clear the screen when we show the menu. If the timeout is zero, we don't. What we do however is read from the kbd to see if any key is in the input queue, which we will process then. It's quite possible that firmwares might initialize the console as soon as one key is read?

If we wouldn't process queued keys you couldn't interact with sd-boot at all anymore, which is quite problematic.

Consider playing around with boot.c and checking what happens if the initial key handling is skipped. If the flicker goes away then then you do have a firmware that does such a "lazy init" of the console framework on kbd read...

Other than that I am not sure what I can propose: the codebase really doesn't clear the screen needlessly, and at least on my machine (thinkpad x1 3rd) boot-up is fully flicker-free the way it is.

**Comment by AdrianVovk** [comment]
@poettering I'll try and get my hands on a different firmware (one of the computers at school) to see if this is a firmware issue.

**Comment by AdrianVovk** [comment]
I just tested on the school computers. The screen doesn't get cleared; there's a full-on modeset

Again, it seems to happen between the firmware and the kernel (plymouth's spinner is not running once the screen turns back on). Firmware cannot be the same, since my computer is a Lenovo and the school's computer is a Dell.

**Comment by poettering** [comment]
are you sure its in sd-boot, and not in the kernel itself?

i mean, look at the sources:

https://github.com/systemd/systemd/blob/master/src/boot/efi/boot.c#L2076

you'll see we do nothing console-related until menu_run() is invoked, which we don't  until we decided to actually show the menu...

the code is not particularly hard to follow

**Comment by AdrianVovk** [comment]
@poettering Yes, In starting to suspect that it's in the kernel. I've done some other hardware tests and it is starting to look like it might be there, though it's really difficult to tell for certain. I'll close the issue here and email the Fedora team for assistance.

Thank you

**Comment by silverhadch** [comment]
@poettering 

I understand that systemd-boot only clears the screen when displaying the menu and, with timeout=0, it doesn’t explicitly do so. However, I’m still seeing an unwanted blank screen between the vendor logo and Plymouth’s BGRT splash. This doesn’t happen with Fedora or Ubuntu’s patched GRUB, even though they also listen for key presses during boot.

From what you mentioned, systemd-boot reads from the keyboard early to check for input, which could trigger a lazy console initialization on some firmware. If that’s what’s causing the flicker, would it be possible to defer or modify how key input is processed to avoid reinitializing the console?

I’d be interested in experimenting with boot.c if there’s a specific area where the initial key handling could be changed or delayed. Could systemd-boot potentially adopt an approach similar to Fedora/Ubuntu’s GRUB, where it listens for key presses without causing a flicker or clearing the screen?
